northbound/query-handler/src/main/java/org/eclipse/sensinact/northbound/query/impl/QueryHandler.java

```diff
@@ -362,14 +376,81 @@ private AbstractResultDTO handleAct(final SensiNactSession userSession, final Qu
      */
     private Collection<ProviderSnapshot> executeFilter(final String filter, final String filterLanguage)
             throws StatusException {
-        final ICriterion parsedFilter = parseFilter(filter, filterLanguage);
+        return executeFilter(parseFilter(filter, filterLanguage));
+    }
+
+    /**
+     * Executes the given parser
+     *
+     * @param filter         Filter
+     * @return Matching snapshot
+     * @throws StatusException Error parsing or executing filter
+     */
+    private Collection<ProviderSnapshot> executeFilter(final ICriterion filter)
+            throws StatusException {
         try {
-            return FilterCommandHelper.executeFilter(gatewayThread, parsedFilter);
+            return FilterCommandHelper.executeFilter(gatewayThread, filter);
         } catch (FilterException e) {
             throw new StatusException(500, "Error executing filter: " + e.getMessage());
         }
     }
 
+    /**
+     * Root of snapshot handling
+     *
+     * @param userSession Caller session
+     * @param query         Query description
+     * @return Result DTO
+     */
+    private AbstractResultDTO handleSnapshot(final SensiNactSession userSession, final QuerySnapshotDTO query)
+            throws Exception {
+        final ResponseSnapshotDTO result = new ResponseSnapshotDTO();
+
+        result.providers = new HashMap<>();
+        result.uri = "/";
+        result.statusCode = 200;
+
+        for (var filter: query.filter) {
+            ICriterion criterion = resourceSelectorFilterFactory.parseResourceSelector(filter);
+            for (var providerSnapshot: executeFilter(criterion)) {
+                if (criterion.getProviderFilter() == null || criterion.getProviderFilter().test(providerSnapshot)) {
```

Comment:
Why are we doing these tests? The results from `executeFilter` should already be filtered correctly.